### PR TITLE
Introduce a container to hold injection strategies

### PR DESCRIPTION
**Pull request by timriley**
This also splits up all the default strategies (args, hash, and kwargs) into their own strategy classes.

@solnic @AMHOL what do you think of this approach? It also lays the foundation for users to pass in their own externally-provided injection strategies, which is a nice piece of flexibility, I think.

I'm using a dry-container to hold the strategies – I like how this gives users the ability to inherit from the default `Dry::AutoInject::Strategies` class to keep all the default strategies and just augment that set with a couple of their own. It may seem a bit funny that this "container agnostic" gem would ship with a dependency on dry-container, but it really is for advanced usage/internal implementation mechanics only.

### CODE REVIEW

**Comment by timriley** [comment]
@flash-gordon as a recent auto_inject contributor, would love to hear your thoughts on this refactor too :)

**Comment by AMHOL** [comment]
@timriley I like this approach, the code is a lot simpler and easier to follow now.

**Comment by flash-gordon** [comment]
@timriley well done! I was curious if we can reduce complexity of nested code which I don't like personally. I tried to move container declaration to a separate module that we already have. Look at my commit here: https://github.com/dry-rb/dry-auto_inject/compare/strategies-container...move-container-definition. Now we don't really need `@container` in Constructor I just left a decision about removing to you. Args constructor still overrides `.new` directly in the class, I didn't change it.

File: dry-auto_inject.gemspec [hunk 1]
```diff
@@ -19,6 +19,8 @@ Gem::Specification.new do |spec|
 
   spec.required_ruby_version = '>= 2.0.0'
 
+  spec.add_runtime_dependency 'dry-container'
```
Comment (solnic) [inline dry-auto_inject.gemspec]: this needs a version spec too

File: lib/dry/auto_inject/builder.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      attr_reader :container
+
+      # @api private
+      attr_reader :strategies
+
+      def initialize(container, options = {})
+        @container = container
+        @strategies = options.fetch(:strategies) { Strategies }
+
+        strategies.keys.each do |strategy_name|
```
Comment (solnic) [inline lib/dry/auto_inject/builder.rb]: `each_key` would be better/faster

**Comment by solnic** [comment]
This looks like a massive improvement :) I had a couple of minor comments only

File: lib/dry/auto_inject/builder.rb (hunk 2 above)
Comment (timriley) [inline lib/dry/auto_inject/builder.rb]: This is a `Dry::Container`, not a regular `Hash`, so all I've got available to me from the public API is `#keys`.

**Comment by timriley** [comment]
@flash-gordon Those changes look good. I'll merge them in! Thanks!

File: lib/dry/auto_inject/builder.rb (hunk 2 above)
Comment (AMHOL) [inline lib/dry/auto_inject/builder.rb]: `each_key` seems like a decent candidate for a method on container, will add it when I get a chance

**Comment by timriley** [comment]
Thanks for the feedback everyone! Pulling this in now.

**Pull request merged**